src/gurt/shm_dict.c

```diff
@@ -0,0 +1,430 @@
+
+	*ht_head = NULL;
+	len_name = strnlen(name, MAX_HT_NAME_LEN);
+	if (len_name >= MAX_HT_NAME_LEN) {
+		DS_ERROR(EINVAL, "hash table name is longer than %d bytes", MAX_HT_NAME_LEN - 1);
+		return EINVAL;
+	}
+
+	/* force bucket size be power of 2 */
+	n_bucket = 1 << bits;
+	shm_mutex_lock(&(d_shm_head->ht_lock), NULL);
+
+	/* loop over existing hash tables to check whether it exists or not */
+	if (d_shm_head->off_ht_head != INVALID_OFFSET) {
+		offset = d_shm_head->off_ht_head;
+		while (offset != INVALID_OFFSET) {
+			ht_head_loc = (struct d_shm_ht_head *)((char *)d_shm_head + offset);
+			if ((strncmp(name, ht_head_loc->ht_name, MAX_HT_NAME_LEN) == 0) &&
+			    (ht_head_loc->n_bucket == n_bucket) &&
+			    (ht_head_loc->n_lock == n_lock)) {
+				/* found existing hash table */
+				*ht_head = ht_head_loc;
+				break;
+			}
+			if (ht_head_loc->next == INVALID_OFFSET) {
+				/* reaching the end of the link list of existing hash tables */
+				*ht_head = NULL;
+				break;
+			}
+			offset = ht_head_loc->next;
+		}
+	}
+
+	if (*ht_head) {
+		/* found existing hash table */
+		shm_mutex_unlock(&(d_shm_head->ht_lock));
+		return 0;
+	}
+
+	/* This hash table does not exist, then create it. */
+	*ht_head = shm_alloc(sizeof(struct d_shm_ht_head) + (sizeof(d_shm_mutex_t) * n_lock) +
+			     (sizeof(long int) * n_bucket));
+	if (*ht_head == NULL)
+		return ENOMEM;
+	ht_head_loc = *ht_head;
+
+	memcpy(ht_head_loc->ht_name, name, len_name + 1);
+	ht_head_loc->n_bucket = n_bucket;
+	ht_head_loc->n_lock   = n_lock;
+
+	p_locks = (d_shm_mutex_t *)((char *)ht_head_loc + sizeof(struct d_shm_ht_head));
+	for (i = 0; i < n_lock; i++) {
+		if (shm_mutex_init(&(p_locks[i])) != 0) {
+			DS_ERROR(errno, "shm_mutex_init() failed");
+			return errno;
+		}
+	}
+	off_next = (long int *)((char *)ht_head_loc + sizeof(struct d_shm_ht_head) +
+				(sizeof(d_shm_mutex_t) * n_lock));
+	for (i = 0; i < n_bucket; i++)
+		off_next[i] = INVALID_OFFSET;
+	/* insert the new hash table header as the first one of the link list */
+	ht_head_loc->prev       = INVALID_OFFSET;
+	ht_head_loc->next       = d_shm_head->off_ht_head;
+	d_shm_head->off_ht_head = (long int)((char *)ht_head_loc - (char *)d_shm_head);
+
+	shm_mutex_unlock(&(d_shm_head->ht_lock));
+	return 0;
+}
+
+bool
+shm_ht_rec_delete(struct d_shm_ht_head *ht_head, const char *key, const int ksize)
+{
+	unsigned int       hash;
+	unsigned int       idx;
+	unsigned int       idx_lock;
+	d_shm_mutex_t     *p_ht_lock;
+	long int           off_next;
+	long int          *p_off_list;
+	struct shm_ht_rec *rec;
+	struct shm_ht_rec *rec_prev = NULL;
+	struct shm_ht_rec *rec_next = NULL;
+
+	hash       = d_hash_string_u32(key, ksize);
+	idx        = hash & (ht_head->n_bucket - 1);
+	idx_lock   = (unsigned int)(idx * ht_head->n_lock * 1.0f / ht_head->n_bucket);
+	p_ht_lock  = (d_shm_mutex_t *)((char *)ht_head + sizeof(struct d_shm_ht_head));
+	p_off_list = (long int *)((char *)p_ht_lock + sizeof(d_shm_mutex_t) * ht_head->n_lock);
```

Comment:
seems to have concurrency issue with `shm_ht_destroy()` function.
The same apply for most of the other functions working on records and `get_shm_ht_with_name()` functions.

Reply:
Thank you very much! I just made some improvements about possible issue of accessing hash table head.  I am working on mitigating the concurrency of hash table record. I will push my commit soon.

Reply:
Could you please review again? Thank you very much!
Now the memory of hash table head and records are not freed as long as the reference count is not zero.